File: src/codegen/io.py
"""
I/O symbol definitions for built-in functions.
"""

from ..utils.nodes import IntType, FloatType, StringType, VoidType
from .utils import FunctionType, Symbol, CName


LIB_NAME = "io"

# Built-in functions for TyC:
# readInt() -> int
# readFloat() -> float
# readString() -> string
# printInt(int) -> void
# printFloat(float) -> void
# printString(string) -> void

IO_SYMBOL_LIST = [
    # Integer I/O
    Symbol("readInt", FunctionType([], IntType()), CName(LIB_NAME)),
    Symbol("printInt", FunctionType([IntType()], VoidType()), CName(LIB_NAME)),
    
    # Float I/O
    Symbol("readFloat", FunctionType([], FloatType()), CName(LIB_NAME)),
    Symbol("printFloat", FunctionType([FloatType()], VoidType()), CName(LIB_NAME)),
    
    # String I/O
    Symbol("readString", FunctionType([], StringType()), CName(LIB_NAME)),
    Symbol("printString", FunctionType([StringType()], VoidType()), CName(LIB_NAME)),
]

File: src/runtime/io.java
import java.util.Scanner;

/**
 * IO class for TyC runtime.
 * Provides input/output operations for TyC programs.
 * 
 * Built-in functions:
 * - int readInt()
 * - float readFloat()
 * - string readString()
 * - void printInt(int value)
 * - void printFloat(float value)
 * - void printString(string value)
 */
public class io {
    private static Scanner scanner = new Scanner(System.in);
    
    // Integer I/O
    public static int readInt() {
        return scanner.nextInt();
    }
    
    public static void printInt(int value) {
        System.out.print(value);
    }
    
    // Float I/O
    public static float readFloat() {
        return scanner.nextFloat();
    }
    
    public static void printFloat(float value) {
        System.out.print(value);
    }
    
    // String I/O
    public static String readString() {
        return scanner.next();
    }
    
    public static void printString(String value) {
        System.out.print(value);
    }
}
